src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,554 @@
+
+        Ok(())
+    }
+
+    fn process_failed_cmd(&mut self, cid: u64, err: Error) {
+        let rctx = self.cmd_ctxs.get_mut(&cid).unwrap();
+
+        match rctx.cmd {
+            Command::Get { ref mut callback, .. } |
+            Command::CommitThenGet { ref mut callback, .. } |
+            Command::RollbackThenGet { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::BatchGet { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Scan { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Prewrite { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Commit { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Cleanup { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Rollback { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+        }
+    }
+
+    fn extract_cmd_context_by_id(&self, cid: u64) -> &Context {
+        let rctx: &RunningCtx = self.cmd_ctxs.get(&cid).unwrap();
+        match rctx.cmd {
+            Command::Get { ref ctx, .. } |
+            Command::BatchGet { ref ctx, .. } |
+            Command::Scan { ref ctx, .. } |
+            Command::Prewrite { ref ctx, .. } |
+            Command::Commit { ref ctx, .. } |
+            Command::CommitThenGet { ref ctx, .. } |
+            Command::Cleanup { ref ctx, .. } |
+            Command::Rollback { ref ctx, .. } |
+            Command::RollbackThenGet { ref ctx, .. } => {
+                ctx
+            }
+        }
+    }
+
+    fn register_report_tick(&self, event_loop: &mut EventLoop<Self>) {
+        if let Err(e) = register_timer(event_loop, Tick::ReportStatistic, REPORT_STATISTIC_INTERVAL) {
+            error!("register report statistic err: {:?}", e);
+        };
+    }
+
+    fn on_report_staticstic_tick(&self, event_loop: &mut EventLoop<Self>) {
+        info!("all running cmd count = {}", self.cmd_ctxs.len());
+
+        self.register_report_tick(event_loop);
+    }
+
+    fn received_new_cmd(&mut self, cmd: Command) {
+        let cid = self.gen_id();
+
+        if cmd.readonly() {
+            // readonly command don't need latch, step to GetSnapshot
+            let ctx: RunningCtx = RunningCtx::new(cid, cmd, vec![]);
+            self.save_cmd_context(cid, ctx);
+
+            if let Err(e) = self.schedch.send(Msg::GetSnapshot{ cid: cid }) {
+                error!("send msg get snapshot failed, cid = {}, error = {:?}", cid, e);
+                self.process_failed_cmd(cid, e);
+                self.finish_cmd(cid);
```

Comment:
I've noticed that every `process_failed_cmd` is followed by a `finish_cmd`. You may abstract a function such as `finish_with_error`.